Improve gateway metadata and add connection information metrics

pcuzner opened the pull request:
This PR improves the gateway metadata information which makes dealing with multiple gateways from prometheus easier. In addition, new metrics showing client connectivity (ceph_nvmeof_host_connection_state) and network status have also been included.

Fixes: #414 

Signed-off-by: Paul Cuzner <pcuzner@ibm.com>

pcuzner committed:
Pass the gateway RPC to the prometheus module

Some stats need to come from the CLI rpc calls, so
this patch provides the means for the exporter to
query the gateway RPC.

Signed-off-by: Paul Cuzner <pcuzner@ibm.com>

pcuzner committed:
Add more metadata and stats to the exporter

This change adds sector size to the bdev metadata,
improves the gateway metadata, adds client connection
state, HA status, load balancing (ANAGRPID), and
includes NIC adapter information for all NICs used by
listeners.

Signed-off-by: Paul Cuzner <pcuzner@ibm.com>

pcuzner commented:
Note once this is merged, I can work on the changes to the ceph-mixins for the alert definitions.

gbregman commented on control/prometheus.py:
```diff
@@ -146,13 +157,27 @@ def _get_spdk_thread_stats(self):
     def _get_subsystems(self):
         return rpc.nvmf.nvmf_get_subsystems(self.spdk_rpc_client)
```
This might be problematic. We experienced some corruptions when there were two SPDK calls in parallel. This is why we introduced an RPC lock in the code and make sure we lock it before calling SPDK. You can replace the call to SPDK here with a call to list_subsystems() using pb2.list_subsystems_req(). This will go through our gRPC and will take the lock.

pcuzner replied:
ACK

pcuzner replied:
Hi @gbregman 
I've hit a little problem..the message format returned drops a lot of the content that the straight spdk rpc call returned, so things like allow_any_host, the listen addresses, hosts etc are no longer visible to the exporter code. To get mostly the same information, I need to execute multiple calls; subsystem list, then listener list -n <nqn> and host list -n <nqn> and finally namespace_list .... this will happen every n seconds - are you ok with that?

Also, max_namespaces is not exposed as far as I can see - I was using that as the basis for an alert :(

gbregman replied:
@pcuzner the latest commit brought back the get_subsystems. So you can now call it via gRPC and get all the information you need.

If you need the max_namespaces, please open an issue and we will add it.

pcuzner commented:
This is the current output with the changes applied

```
# HELP python_info Python platform information
# TYPE python_info gauge
python_info{implementation="CPython",major="3",minor="9",patchlevel="18",version="3.9.18"} 1.0
# HELP process_virtual_memory_bytes Virtual memory size in bytes.
# TYPE process_virtual_memory_bytes gauge
process_virtual_memory_bytes 1.708568576e+09
# HELP process_resident_memory_bytes Resident memory size in bytes.
# TYPE process_resident_memory_bytes gauge
process_resident_memory_bytes 5.9953152e+07
# HELP process_start_time_seconds Start time of the process since unix epoch in seconds.
# TYPE process_start_time_seconds gauge
process_start_time_seconds 1.70736480523e+09
# HELP process_cpu_seconds_total Total user and system CPU time spent in seconds.
# TYPE process_cpu_seconds_total counter
process_cpu_seconds_total 0.6000000000000001
# HELP process_open_fds Number of open file descriptors.
# TYPE process_open_fds gauge
process_open_fds 24.0
# HELP process_max_fds Maximum number of open file descriptors.
# TYPE process_max_fds gauge
process_max_fds 20480.0
# HELP ceph_nvmeof_gateway_info Gateway information
# TYPE ceph_nvmeof_gateway_info gauge
ceph_nvmeof_gateway_info{addr="0.0.0.0",hostname="c742790cb1d4",name="c742790cb1d4",port="5500",spdk_version="23.01.1",version="1.0.0"} 1.0
# HELP ceph_nvmeof_bdev_capacity_bytes BDEV Capacity
# TYPE ceph_nvmeof_bdev_capacity_bytes gauge
ceph_nvmeof_bdev_capacity_bytes{bdev_name="testdisk"} 5.36870912e+09
# HELP ceph_nvmeof_bdev_metadata BDEV Metadata
# TYPE ceph_nvmeof_bdev_metadata gauge
ceph_nvmeof_bdev_metadata{bdev_name="testdisk",block_size="512",namespace="",pool_name="rbd",rbd_name="testdisk"} 1.0
# HELP ceph_nvmeof_bdev_reads_completed_total Total number of read operations completed
# TYPE ceph_nvmeof_bdev_reads_completed_total counter
ceph_nvmeof_bdev_reads_completed_total{bdev_name="testdisk"} 2.0
# HELP ceph_nvmeof_bdev_writes_completed_total Total number of write operations completed
# TYPE ceph_nvmeof_bdev_writes_completed_total counter
ceph_nvmeof_bdev_writes_completed_total{bdev_name="testdisk"} 0.0
# HELP ceph_nvmeof_bdev_read_bytes_total Total number of bytes read successfully
# TYPE ceph_nvmeof_bdev_read_bytes_total counter
ceph_nvmeof_bdev_read_bytes_total{bdev_name="testdisk"} 36864.0
# HELP ceph_nvmeof_bdev_written_bytes_total Total number of bytes written successfully
# TYPE ceph_nvmeof_bdev_written_bytes_total counter
ceph_nvmeof_bdev_written_bytes_total{bdev_name="testdisk"} 0.0
# HELP ceph_nvmeof_bdev_read_seconds_total Total time spent servicing READ I/O
# TYPE ceph_nvmeof_bdev_read_seconds_total counter
ceph_nvmeof_bdev_read_seconds_total{bdev_name="testdisk"} 0.0005702764285714285
# HELP ceph_nvmeof_bdev_write_seconds_total Total time spent servicing WRITE I/O
# TYPE ceph_nvmeof_bdev_write_seconds_total counter
ceph_nvmeof_bdev_write_seconds_total{bdev_name="testdisk"} 0.0
# HELP ceph_nvmeof_reactor_seconds_total time reactor thread active with I/O
# TYPE ceph_nvmeof_reactor_seconds_total counter
ceph_nvmeof_reactor_seconds_total{mode="busy",name="nvmf_tgt_poll_group_0"} 0.009171436071428571
ceph_nvmeof_reactor_seconds_total{mode="idle",name="nvmf_tgt_poll_group_0"} 189.03610959357144
# HELP ceph_nvmeof_subsystem_metadata Metadata describing the subsystem configuration
# TYPE ceph_nvmeof_subsystem_metadata gauge
ceph_nvmeof_subsystem_metadata{allow_any_host="no",ha_enabled="no",model_number="SPDK bdev Controller",nqn="nqn.2016-06.io.spdk:cnode1",serial_number="SPDK60804759770178"} 1.0
# HELP ceph_nvmeof_subsystem_listener_count Number of listener addresses used by the subsystem
# TYPE ceph_nvmeof_subsystem_listener_count gauge
ceph_nvmeof_subsystem_listener_count{nqn="nqn.2016-06.io.spdk:cnode1"} 1.0
# HELP ceph_nvmeof_subsystem_host_count Number of hosts defined to the subsystem
# TYPE ceph_nvmeof_subsystem_host_count gauge
ceph_nvmeof_subsystem_host_count{nqn="nqn.2016-06.io.spdk:cnode1"} 1.0
# HELP ceph_nvmeof_subsystem_namespace_metadata Namespace information for the subsystem
# TYPE ceph_nvmeof_subsystem_namespace_metadata gauge
ceph_nvmeof_subsystem_namespace_metadata{bdev_name="testdisk",load_balancing_group="0",nqn="nqn.2016-06.io.spdk:cnode1",nsid="1"} 1.0
# HELP ceph_nvmeof_host_connection_state Host connection state 0=disconnected, 1= connected
# TYPE ceph_nvmeof_host_connection_state gauge
ceph_nvmeof_host_connection_state{gw_name="c742790cb1d4",host_addr="<n/a>",host_nqn="nqn.2014-08.org.nvmexpress:uuid:ee889718-8c69-40d3-8e78-5be049f966a6",nqn="nqn.2016-06.io.spdk:cnode1"} 0.0
# HELP ceph_nvmeof_rpc_method_seconds Run times of the RPC method calls
# TYPE ceph_nvmeof_rpc_method_seconds gauge
ceph_nvmeof_rpc_method_seconds{method="_get_bdev_info"} 0.007887840270996094
ceph_nvmeof_rpc_method_seconds{method="_get_bdev_io_stats"} 0.007985115051269531
ceph_nvmeof_rpc_method_seconds{method="_get_spdk_thread_stats"} 0.008033990859985352
ceph_nvmeof_rpc_method_seconds{method="_get_subsystems"} 0.008391380310058594
ceph_nvmeof_rpc_method_seconds{method="_get_connection_map"} 0.023740053176879883
ceph_nvmeof_rpc_method_seconds{method="_get_host_info"} 0.007892370223999023
ceph_nvmeof_rpc_method_seconds{method="_get_listener_info"} 0.008116483688354492
ceph_nvmeof_rpc_method_seconds{method="_get_namespace_info"} 0.015845537185668945
```

pcuzner commented:
@gbregman I notice that now the get_subsystems calls requires a context object. I couldn't find the implementation of the context object and have been using pb2.req_status() but what should I be using?

pcuzner committed:
Add classes to handle NIC configuration

Two classes are added - NICS and NIC. These are helper
classes to create a view of the hosts network configuration.

Uses the netifaces package to enumerate the ip addresses
and available NICs

Signed-off-by: Paul Cuzner <pcuzner@ibm.com>

pcuzner committed:
Update python dependencies for netifaces

Signed-off-by: Paul Cuzner <pcuzner@ibm.com>

pcuzner commented:
@gbregman The latest changes switch to the get_subsystems call, and also add in NIC state metrics so we can potentially alert when a NIC goes down that is associated with a subsystem listener.

pcuzner commented:
This is the latest output. You can see the operstate of the NIC with ceph_nvmeof_subsystem_listener_iface_info and the NICs speed via ceph_nvmeof_subsystem_listener_iface_speed_bytes 
```
# HELP python_info Python platform information
# TYPE python_info gauge
python_info{implementation="CPython",major="3",minor="9",patchlevel="18",version="3.9.18"} 1.0
# HELP process_virtual_memory_bytes Virtual memory size in bytes.
# TYPE process_virtual_memory_bytes gauge
process_virtual_memory_bytes 1.715499008e+09
# HELP process_resident_memory_bytes Resident memory size in bytes.
# TYPE process_resident_memory_bytes gauge
process_resident_memory_bytes 6.2144512e+07
# HELP process_start_time_seconds Start time of the process since unix epoch in seconds.
# TYPE process_start_time_seconds gauge
process_start_time_seconds 1.70802978776e+09
# HELP process_cpu_seconds_total Total user and system CPU time spent in seconds.
# TYPE process_cpu_seconds_total counter
process_cpu_seconds_total 5.140000000000001
# HELP process_open_fds Number of open file descriptors.
# TYPE process_open_fds gauge
process_open_fds 24.0
# HELP process_max_fds Maximum number of open file descriptors.
# TYPE process_max_fds gauge
process_max_fds 20480.0
# HELP ceph_nvmeof_gateway_info Gateway information
# TYPE ceph_nvmeof_gateway_info gauge
ceph_nvmeof_gateway_info{addr="0.0.0.0",group="group-1",hostname="24117969b875",name="24117969b875",port="5500",spdk_version="23.01.1",version="1.0.0"} 1.0
# HELP ceph_nvmeof_bdev_capacity_bytes BDEV Capacity
# TYPE ceph_nvmeof_bdev_capacity_bytes gauge
ceph_nvmeof_bdev_capacity_bytes{bdev_name="bdev_7516ba03-d9fe-49b4-b14c-476a25e7f3ea"} 1.073741824e+09
# HELP ceph_nvmeof_bdev_metadata BDEV Metadata
# TYPE ceph_nvmeof_bdev_metadata gauge
ceph_nvmeof_bdev_metadata{bdev_name="bdev_7516ba03-d9fe-49b4-b14c-476a25e7f3ea",block_size="512",namespace="",pool_name="rbd",rbd_name="testdisk"} 1.0
# HELP ceph_nvmeof_bdev_reads_completed_total Total number of read operations completed
# TYPE ceph_nvmeof_bdev_reads_completed_total counter
ceph_nvmeof_bdev_reads_completed_total{bdev_name="bdev_7516ba03-d9fe-49b4-b14c-476a25e7f3ea"} 52.0
# HELP ceph_nvmeof_bdev_writes_completed_total Total number of write operations completed
# TYPE ceph_nvmeof_bdev_writes_completed_total counter
ceph_nvmeof_bdev_writes_completed_total{bdev_name="bdev_7516ba03-d9fe-49b4-b14c-476a25e7f3ea"} 0.0
# HELP ceph_nvmeof_bdev_read_bytes_total Total number of bytes read successfully
# TYPE ceph_nvmeof_bdev_read_bytes_total counter
ceph_nvmeof_bdev_read_bytes_total{bdev_name="bdev_7516ba03-d9fe-49b4-b14c-476a25e7f3ea"} 1.110016e+06
# HELP ceph_nvmeof_bdev_written_bytes_total Total number of bytes written successfully
# TYPE ceph_nvmeof_bdev_written_bytes_total counter
ceph_nvmeof_bdev_written_bytes_total{bdev_name="bdev_7516ba03-d9fe-49b4-b14c-476a25e7f3ea"} 0.0
# HELP ceph_nvmeof_bdev_read_seconds_total Total time spent servicing READ I/O
# TYPE ceph_nvmeof_bdev_read_seconds_total counter
ceph_nvmeof_bdev_read_seconds_total{bdev_name="bdev_7516ba03-d9fe-49b4-b14c-476a25e7f3ea"} 0.01712989892857143
# HELP ceph_nvmeof_bdev_write_seconds_total Total time spent servicing WRITE I/O
# TYPE ceph_nvmeof_bdev_write_seconds_total counter
ceph_nvmeof_bdev_write_seconds_total{bdev_name="bdev_7516ba03-d9fe-49b4-b14c-476a25e7f3ea"} 0.0
# HELP ceph_nvmeof_reactor_seconds_total time reactor thread active with I/O
# TYPE ceph_nvmeof_reactor_seconds_total counter
ceph_nvmeof_reactor_seconds_total{mode="busy",name="nvmf_tgt_poll_group_0"} 0.03954470964285714
ceph_nvmeof_reactor_seconds_total{mode="idle",name="nvmf_tgt_poll_group_0"} 1908.3708052260715
# HELP ceph_nvmeof_subsystem_metadata Metadata describing the subsystem configuration
# TYPE ceph_nvmeof_subsystem_metadata gauge
ceph_nvmeof_subsystem_metadata{allow_any_host="no",group="group-1",ha_enabled="no",model_number="SPDK bdev Controller",nqn="nqn.2016-06.io.spdk:cnode1",serial_number="SPDK65724372656387"} 1.0
# HELP ceph_nvmeof_subsystem_listener_count Number of listener addresses used by the subsystem
# TYPE ceph_nvmeof_subsystem_listener_count gauge
ceph_nvmeof_subsystem_listener_count{nqn="nqn.2016-06.io.spdk:cnode1"} 1.0
# HELP ceph_nvmeof_subsystem_host_count Number of hosts defined to the subsystem
# TYPE ceph_nvmeof_subsystem_host_count gauge
ceph_nvmeof_subsystem_host_count{nqn="nqn.2016-06.io.spdk:cnode1"} 1.0
# HELP ceph_nvmeof_subsystem_namespace_limit Maximum namespaces supported
# TYPE ceph_nvmeof_subsystem_namespace_limit gauge
ceph_nvmeof_subsystem_namespace_limit{nqn="nqn.2016-06.io.spdk:cnode1"} 256.0
# HELP ceph_nvmeof_subsystem_namespace_metadata Namespace information for the subsystem
# TYPE ceph_nvmeof_subsystem_namespace_metadata gauge
ceph_nvmeof_subsystem_namespace_metadata{anagrpid="1",bdev_name="bdev_7516ba03-d9fe-49b4-b14c-476a25e7f3ea",nqn="nqn.2016-06.io.spdk:cnode1",nsid="1"} 1.0
# HELP ceph_nvmeof_host_connection_state Host connection state 0=disconnected, 1=connected
# TYPE ceph_nvmeof_host_connection_state gauge
ceph_nvmeof_host_connection_state{gw_name="24117969b875",host_addr="192.168.13.3:4420",host_nqn="nqn.2014-08.org.nvmexpress:uuid:ee889718-8c69-40d3-8e78-5be049f966a6",nqn="nqn.2016-06.io.spdk:cnode1"} 1.0
# HELP ceph_nvmeof_subsystem_listener_iface_info Interface information
# TYPE ceph_nvmeof_subsystem_listener_iface_info gauge
ceph_nvmeof_subsystem_listener_iface_info{device="eth0",duplex="full",mac_address="02:42:c0:a8:0d:03",operstate="up"} 1.0
# HELP ceph_nvmeof_subsystem_listener_iface_speed_bytes Link speed of the Listener interface
# TYPE ceph_nvmeof_subsystem_listener_iface_speed_bytes gauge
ceph_nvmeof_subsystem_listener_iface_speed_bytes{device="eth0"} 10000.0
# HELP ceph_nvmeof_subsystem_listener_iface_nqn_info Subsystem usage of a NIC device
# TYPE ceph_nvmeof_subsystem_listener_iface_nqn_info gauge
ceph_nvmeof_subsystem_listener_iface_nqn_info{device="eth0",nqn="nqn.2016-06.io.spdk:cnode1"} 1.0
# HELP ceph_nvmeof_rpc_method_seconds Run times of the RPC method calls
# TYPE ceph_nvmeof_rpc_method_seconds gauge
ceph_nvmeof_rpc_method_seconds{method="_get_bdev_info"} 0.005443572998046875
ceph_nvmeof_rpc_method_seconds{method="_get_bdev_io_stats"} 0.008162736892700195
ceph_nvmeof_rpc_method_seconds{method="_get_spdk_thread_stats"} 0.007859468460083008
ceph_nvmeof_rpc_method_seconds{method="_get_subsystems"} 0.00840449333190918
ceph_nvmeof_rpc_method_seconds{method="_list_subsystems"} 0.007885217666625977
ceph_nvmeof_rpc_method_seconds{method="_get_connection_map"} 0.024187326431274414

```

gbregman commented:
> @gbregman I notice that now the get_subsystems calls requires a context object. I couldn't find the implementation of the context object and have been using pb2.req_status() but what should I be using?

The context parameter is passed by the infrastructure as part of gRPC. As you just read the value and don't make any changes you can just use None as the value.

baum commented:
@pcuzner lgtm 🙏

baum commented:
@pcuzner could you please rebase the change, to include the  new required tests, and merge please? Thank you 🖖

pcuzner commented:
> @pcuzner could you please rebase the change, to include the new required tests, and merge please? Thank you 🖖
@baum rebased and pushed.

pcuzner commented:
Just looking at #487. I'll look at that issue and potentially put the patches required in this PR.

pcuzner committed:
Prometheus module updated to take rpc locks

Locks needed on the rpc calls to avoid race condition
when the prometheus module starts issuing requests
before the server has fully initiallized and applied the
current configuration.

Signed-off-by: Paul Cuzner <pcuzner@ibm.com>

pcuzner commented:
@caroav please test this PR through automation.

pcuzner committed:
dependencies update

Signed-off-by: Paul Cuzner <pcuzner@ibm.com>

pcuzner commented:
Had to push again due to merge of #489
